### PR TITLE
codeintel: Cache github authz responses

### PR DESCRIPTION
**Pull request by efritz**
Previously, we would take the user-supplied github token on every LSIF upload and hit the GitHub API directly to see if the user has authorship over the target repo.

This PR adds a caching layer (via rcache, in Redis) for those responses for a particular github token/repository pair. Before serializing into the redis cache, we take a sum of the data so that we do not store user github tokens anywhere in our stack.

This should reduce the number of requests we make to github, especially when we have users (like ourselves) with high commit frequency and lots of upload jobs launching in quick sequence.

### CODE REVIEW

File: enterprise/cmd/frontend/internal/codeintel/httpapi/auth_github_cache.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+		return false, false
+	}
+
+	err := json.Unmarshal(b, &authorized)
+	return authorized, err == nil
+}
+
+func (c *GitHubAuthCache) Set(key string, authorized bool) {
+	b, _ := json.Marshal(authorized)
+	c.cache.Set(key, b)
```
Comment (eseliger) [inline enterprise/cmd/frontend/internal/codeintel/httpapi/auth_github_cache.go:31]: why can't this fail 🧐

**Review by andreeleuterio — APPROVED** [review]

**Review by eseliger — APPROVED** [review]
should we also remap the err code?

File: enterprise/cmd/frontend/internal/codeintel/httpapi/auth_github_cache.go (hunk 1 above)
Comment (efritz) [reply enterprise/cmd/frontend/internal/codeintel/httpapi/auth_github_cache.go:31]: rcache just logs it internally

File: enterprise/cmd/frontend/internal/codeintel/httpapi/auth_github.go [hunk 2]
```diff
@@ -19,12 +19,40 @@ var (
...
+
+		return 0, nil
+	}
+
+	defer func() {
+		switch err {
+		case nil:
+			githubAuthCache.Set(key, true)
+		case ErrGitHubUnauthorized:
+			// githubAuthCache.Set(key, false)
```
Comment (eseliger) [inline enterprise/cmd/frontend/internal/codeintel/httpapi/auth_github.go:43]:
```suggestion
```

**Review by eseliger — APPROVED** [review]

File: enterprise/cmd/frontend/internal/codeintel/httpapi/auth_github.go [hunk 3]
```diff
@@ -19,12 +19,39 @@ var (
...
+
+	defer func() {
+		switch err {
+		case nil:
+			githubAuthCache.Set(key, true)
+		case ErrGitHubUnauthorized:
+			// Note: We explicitly do not store false here in case a user is
+			// adjusting permissions on a cache key. Storing false here would
+			// result in a cached rejection after the key has been modified
+			// on the code host.
```
Comment (Strum355) [inline enterprise/cmd/frontend/internal/codeintel/httpapi/auth_github.go:46]: TIL a key can be modified. Always created a new one personally

**Review by Strum355 — APPROVED** [review]

File: enterprise/cmd/frontend/internal/codeintel/httpapi/auth_github.go (hunk 3 above)
Comment (efritz) [reply enterprise/cmd/frontend/internal/codeintel/httpapi/auth_github.go:46]: You can authorize a user for a new repo and the key should reflect that.

**Pull request merged**